### PR TITLE
New Scripts: IncidentFields and GetFieldsByIncidentType.

### PR DESCRIPTION
**Pull request by tundisto**
This PR is for new two automations related to incident fields:

IncidentFields: returns a dict of all incident fields in the system

GetFieldsByIncidentType:  returns the incident fields associated with a given incident type

## Status
Ready - PR https://github.com/demisto/content/pull/5275 has been merged now

## Related Issues
N/A

## Description
IncidentFields: returns a dict of all incident fields in the system

GetFieldsByIncidentType:  returns the incident fields associated with a given incident type

## Screenshots
Paste here any images that will help the reviewer

## Required version of Demisto
4.x/5.x

## Does it break backward compatibility?
   - No

## Must have
- [X] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

## Dependencies
Mention the dependencies of the entity you changed as given from the precommit hooks in checkboxes, and tick after tested them.
- [ ] Dependency 1
- [ ] Dependency 2
- [ ] Dependency 3

## Additional changes
Depends on new CommonServerPython argToBoolean() function in PR https://github.com/demisto/content/pull/5275 

## Technical writer review
Mention and link to the files that require a technical writer review.
- [ ] [YAML file](link)
- [ ] [CHANGELOG](link)

### CODE REVIEW

**Comment by ronykoz** [comment]
@tundisto i saw that the status is on hold, please let me know when can we start reviewing this

**Comment by tundisto** [comment]
@ronykoz you're welcome to start reviewing now, if you'd like.  It's just that CI builds will continue to fail until CommonServerPython gets updated after the other PR gets merged into master.

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    explicit_only = False
+    if 'explicit_only' in args and argToBoolean(args['explicit_only']) is True:
+        explicit_only = True
+
+    # generate results
+    types = []
+    if exclude_system is True:
+        # only return non-system fields
+        for field in incident_fields:  # using multiple if statements for readability
+            if field['system'] is False:  # exclude system fields
```
Comment (guykeller) [inline Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: isn't that already checked using exclude_system?

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                if field['associatedToAll'] is True and explicit_only is False:
+                    # if explicit_only is false, include fields associated to all incident types
+                    types.append(field[name_key])
+                elif field['associatedTypes'] is not None and incident_type in field['associatedTypes']:
```
Comment (guykeller) [inline Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: let's save each condition in a variable with an indicative name, and combine the two if statements to one `if X or Y` statement

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+def main():
+    # get incident fields
+    res = demisto.executeCommand('demisto-api-get', {'uri': '/incidentfields'})
+    if is_error(res):
+        return_error(res[0]['Contents'])
+
+    fields = res[0]['Contents']['response']
+
+    # 'fields' contains non-incident fields, as well, so let's make a version containing only incident fields
+    incident_fields = [field for field in fields if field['id'].startswith('incident_')]
```
Comment (guykeller) [inline Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: Why aren't fields that don't start with `_incident` returned?

File: Scripts/IncidentFields/script-IncidentFields.yml [hunk 4]
```diff
@@ -0,0 +1,107 @@
+commonfields:
+  id: 0fc0b044-47fe-4d52-8a90-6e46a65c0389
+  version: 192
+name: IncidentFields
+script: |
```
Comment (guykeller) [inline Scripts/IncidentFields/script-IncidentFields.yml]: scripts should be packagified and this file should not exist

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        if field['name'] == name:
+            return field
+    return None
+
+
+def get_field_by_short_name(fields, name):
+    for field in fields:
+        if field['cliName'] == name:
+            return field
+    return None
```
Comment (guykeller) [inline Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: seems these functions are never used

File: Scripts/IncidentFields/IncidentFields.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        if field['name'] == name:
+            return field
+    return None
+
+
+def get_field_by_short_name(fields, name):
+    for field in fields:
+        if field['cliName'] == name:
+            return field
+    return None
```
Comment (guykeller) [inline Scripts/IncidentFields/IncidentFields.py]: same here, these functions are never used

File: Scripts/IncidentFields/IncidentFields.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+
+def get_short_name(value):
+    return re.sub(r'[^A-Za-z0-9]', '', value).lower()
```
Comment (guykeller) [inline Scripts/IncidentFields/IncidentFields.py]: this function isn't used as well

File: Scripts/IncidentFields/IncidentFields.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+    fields = res[0]['Contents']['response']
+
+    # 'fields' contains non-incident fields, as well, so let's make a version containing only incident fields
+    incident_fields = [field for field in fields if field['id'].startswith('incident_')]
+
+    # get arguments
+    args = demisto.args()
+
+    custom_only = False
```
Comment (guykeller) [inline Scripts/IncidentFields/IncidentFields.py]: this name isn't clear, what does this condition stand for?

File: Scripts/IncidentFields/IncidentFields.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    if 'custom' in args and argToBoolean(args['custom']) is True:
+        custom_only = True
```
Comment (guykeller) [inline Scripts/IncidentFields/IncidentFields.py]: let's change the name of this argument to something more readable, like include_system_fields (if I understood what this stands for correctly).

File: Scripts/IncidentFields/IncidentFields.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+    short_names = False
+    if 'short_names' in args and argToBoolean(args['short_names']) is True:
+        short_names = True
```
Comment (guykeller) [inline Scripts/IncidentFields/IncidentFields.py]: Let's rename to something more readable.
Like use_cli_name, or use_short_name if that's the terminology the users know

File: Scripts/IncidentFields/IncidentFields.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    for field in incident_fields:
+
+        if custom_only is True and field['system'] is True:
+            continue
+
+        field_name = field['name']
+        if short_names is True:
+            field_name = field['cliName']
+
+        output_field = {
```
Comment (guykeller) [inline Scripts/IncidentFields/IncidentFields.py]:
```suggestion
        field_context_data = {
```

File: Scripts/IncidentFields/IncidentFields.py [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+            'associatedToAll': field['associatedToAll']
+        }
+
+        if output_field['associatedToAll'] is False:
+            if len(field['associatedTypes']) != 0:
+                output_field['associatedTypes'] = field['associatedTypes']
+            else:
+                output_field['associatedTypes'] = None
+        else:
+            output_field['associatedTypes'] = 'all'
```
Comment (guykeller) [inline Scripts/IncidentFields/IncidentFields.py]: let's move this to an external function called `build_field_context_data` or something like that

File: Scripts/IncidentFields/IncidentFields.py [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    custom_only = False
+    if 'custom' in args and argToBoolean(args['custom']) is True:
+        custom_only = True
+
+    short_names = False
+    if 'short_names' in args and argToBoolean(args['short_names']) is True:
+        short_names = True
+
+    # build a dict of all field types and their associated case types
+    types = {}
```
Comment (guykeller) [inline Scripts/IncidentFields/IncidentFields.py]: Why is it names `types`? seems to contain fields data

File: Scripts/IncidentFields/IncidentFields.py [hunk 14]
```diff
@@ -0,0 +1,77 @@
+import demistomock as demisto
+from CommonServerPython import *
+
```
Comment (guykeller) [inline Scripts/IncidentFields/IncidentFields.py]: Please apply relevant notes from the GetFieldsByIncidentType file here as well

File: Scripts/GetFieldsByIncidentType/script-GetFieldsByIncidentType.yml [hunk 15]
```diff
@@ -0,0 +1,110 @@
+commonfields:
+  id: fb45fb17-eaf7-435f-874d-f4fa43b4c58a
+  version: 108
+name: GetFieldsByIncidentType
+script: |
```
[the remark this answers is not in the document]
Comment (tundisto) [reply Scripts/GetFieldsByIncidentType/script-GetFieldsByIncidentType.yml]: True enough.  Forgot to delete the file.

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py [hunk 16]
```diff
@@ -0,0 +1,79 @@
+import demistomock as demisto
+from CommonServerPython import *
+from pprint import pformat
+
+
+def get_field_by_long_name(fields, name):
```
[the remark this answers is not in the document]
Comment (tundisto) [reply Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: Function is leftover from an old version and is now removed

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    for field in fields:
+        if field['name'] == name:
+            return field
+    return None
+
+
+def get_field_by_short_name(fields, name):
```
[the remark this answers is not in the document]
Comment (tundisto) [reply Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: Function is leftover from an old version and is now removed

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+    # 'fields' contains non-incident fields, as well, so let's make a version containing only incident fields
+    incident_fields = [field for field in fields if field['id'].startswith('incident_')]
+
+    # get arguments
+    args = demisto.args()
+
+    incident_type = args['incident_type']
+
+    exclude_system = False
+    if 'custom' in args and argToBoolean(args['custom']) is True:
```
[the remark this answers is not in the document]
Comment (tundisto) [reply Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: I'm well acquainted about the convention; I wrote it that way for readabilty.  Will change it as you object

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+        exclude_system = True
```
[the remark this answers is not in the document]
Comment (tundisto) [reply Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: Quite right, I changed the param name and didn't update all the right places.

File: Scripts/IncidentFields/IncidentFields.py (hunk 13 above)
Comment (tundisto) [reply Scripts/IncidentFields/IncidentFields.py]: Renamed it to matched_fields

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+    if 'short_names' in args and argToBoolean(args['short_names']) is True:
+        name_key = 'cliName'
+
+    explicit_only = False
+    if 'explicit_only' in args and argToBoolean(args['explicit_only']) is True:
+        explicit_only = True
+
+    # generate results
+    types = []
+    if exclude_system is True:
```
[the remark this answers is not in the document]
Comment (tundisto) [reply Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: Ok

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+    if 'custom' in args and argToBoolean(args['custom']) is True:
+        exclude_system = True
+
+    name_key = 'name'
+    if 'short_names' in args and argToBoolean(args['short_names']) is True:
+        name_key = 'cliName'
+
+    explicit_only = False
+    if 'explicit_only' in args and argToBoolean(args['explicit_only']) is True:
+        explicit_only = True
```
[the remark this answers is not in the document]
Comment (tundisto) [reply Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: It means that if the field has 'associate to all' checked, it will be excluded from the results.  I've updated the description to read "If  true, only fields which are explicitly associated to the incident type will be included (i.e. excludes fields with the "associate to any" box ticked)"

Really?  I think the args.get() convention is less readable than the way I wrote it.

Also, doesn't args.get() fail to normalise strings into booleans?  That's why I pass it through argToBoolean().

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tundisto) [reply Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: When I'm looking for a null value, I like to explicitly say that's what I'm looking for, as opposed to other truthy values, as it says something about what's being evaluated.

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+
+    else:
+        # return all fields
+        for field in incident_fields:  # using multiple if statements for readability
+            if field['associatedToAll'] is True and explicit_only is False:
+                # if explicit_only is false, include fields associated to all incident types
+                types.append(field[name_key])
+            elif field['associatedTypes'] is not None and incident_type in field['associatedTypes']:
+                # include fields where incident type is in associatedTypes
+                types.append(field[name_key])
```
[the remark this answers is not in the document]
Comment (tundisto) [reply Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: Done

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py (hunk 1 above)
Comment (tundisto) [reply Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: exclude_system excludes fields that have the system attribute enabled, which is a number of OOTB fields.  `if field['system'] is False` enforces exclude_system.

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py (hunk 3 above)
Comment (tundisto) [reply Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: The comment above it explains it

File: Scripts/IncidentFields/script-IncidentFields.yml (hunk 4 above)
Comment (tundisto) [reply Scripts/IncidentFields/script-IncidentFields.yml]: Yep, forgot to delete it.  It's gone now.

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py (hunk 5 above)
Comment (tundisto) [reply Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: Yep, leftover from an old version and now removed

File: Scripts/IncidentFields/IncidentFields.py (hunk 6 above)
Comment (tundisto) [reply Scripts/IncidentFields/IncidentFields.py]: Yep, leftover from an old version and now removed

File: Scripts/IncidentFields/IncidentFields.py (hunk 7 above)
Comment (tundisto) [reply Scripts/IncidentFields/IncidentFields.py]: Yep, leftover from an old version and now removed

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py (hunk 2 above)
Comment (tundisto) [reply Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: Sorry, why?

File: Scripts/IncidentFields/IncidentFields.py (hunk 11 above)
Comment (tundisto) [reply Scripts/IncidentFields/IncidentFields.py]: ok

File: Scripts/IncidentFields/IncidentFields.py (hunk 12 above)
Comment (tundisto) [reply Scripts/IncidentFields/IncidentFields.py]: done

File: Scripts/IncidentFields/IncidentFields.py (hunk 8 above)
Comment (tundisto) [reply Scripts/IncidentFields/IncidentFields.py]: Quite right, I changed the param name and didn't update all the right places.  Is now explicit_only rather than explicit_only.

File: Scripts/IncidentFields/IncidentFields.py (hunk 10 above)
Comment (tundisto) [reply Scripts/IncidentFields/IncidentFields.py]: 'cliName' is the property name that the `/incidentfields` API call gives to fields' shortnames, and I think is less descriptive, as what does it have to do with the CLI?

Not really sure what term is that users use (I don't think there is anything on the support site referencing it), but "short names" is what I call the shortened version of incident field names.  I'm happy to leave it as-is.

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (tundisto) [reply Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: Ok

File: Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tundisto) [reply Scripts/GetFieldsByIncidentType/GetFieldsByIncidentType.py]: A reason I use the 'is True|False' is that since this is referring to dict property, a string or numeric value other than zero would also evaluate true, which I don't want.  Using `is True` is more appropriate as it enforces boolean type checking on the property.

This is also why I like using `is True|False` rather than `if x` or `if not x` -- I want to fail right away if a non-boolean type value somehow creeps its way into the data.  It's also a micro-optimisation in that all it has to do, in theory, is a reference comparison.

To sum it up, I like type checking.

**Comment by anara123** [comment]
@guykeller what is the status here? 
seems that the PR is ready to move to documentation review step

**Comment by guykeller** [comment]
@anara123 There was a problem with the base branch that stopped us from moving forward, which was fixed this evening

**Comment by kirbles19** [comment]
@tundisto @ronykoz Done.
